metronome/rocksdb/src/io/iohk/metronome/rocksdb/RocksDBStore.scala

```diff
@@ -0,0 +1,374 @@
+package io.iohk.metronome.rocksdb
+
+import cats._
+import cats.implicits._
+import cats.data.ReaderT
+import cats.effect.{Resource, Sync}
+import cats.free.Free.liftF
+import io.iohk.metronome.storage.{KVStore, KVStoreOp}
+import io.iohk.metronome.storage.KVStoreOp.{Put, Get, Delete}
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import org.rocksdb.{
+  RocksDB,
+  WriteBatch,
+  WriteOptions,
+  ReadOptions,
+  Options,
+  DBOptions,
+  ColumnFamilyOptions,
+  ColumnFamilyDescriptor,
+  ColumnFamilyHandle,
+  BlockBasedTableConfig,
+  BloomFilter,
+  CompressionType,
+  ClockCache
+}
+import scodec.Codec
+import scodec.bits.BitVector
+import scala.collection.mutable
+import java.nio.file.Path
+import scala.annotation.nowarn
+
+class RocksDBStore[F[_]: Sync](
+    db: RocksDB,
+    readOptions: ReadOptions,
+    rwlock: ReentrantReadWriteLock,
+    handles: Map[RocksDBStore.Namespace, ColumnFamilyHandle]
+) {
+  import RocksDBStore.{Namespace, DBQuery, autoCloseableR}
+
+  // Batch execution needs these variables for accumulating operations
+  // and executing them against the database. They are going to be
+  // passed along in a Reader monad to the Free compiler.
+  type BatchEnv = (WriteOptions, WriteBatch)
+  // Type aliases to support the `~>` transformation with types that
+  // only have 1 generic type argument `A`.
+  type Batch[A]          = ({ type L[A] = ReaderT[F, BatchEnv, A] })#L[A]
+  type KVNamespacedOp[A] = ({ type L[A] = KVStoreOp[Namespace, A] })#L[A]
+
+  // Batches can interleave multiple reads (and writes);
+  // to make sure they see a consistent view, writes are
+  // isolated from reads via locks, so for example if we
+  // read an ID, then retrieve the record from a different
+  // collection, we can be sure it hasn't been deleted in
+  // between the two operations.
+  private val lockRead    = Sync[F].delay(rwlock.readLock().lock())
+  private val unlockRead  = Sync[F].delay(rwlock.readLock().unlock())
+  private val lockWrite   = Sync[F].delay(rwlock.writeLock().lock())
+  private val unlockWrite = Sync[F].delay(rwlock.writeLock().unlock())
+
+  private def withReadLock[A](fa: F[A]): F[A] =
+    Sync[F].bracket(lockRead)(_ => fa)(_ => unlockRead)
+
+  private def withWriteLock[A](fa: F[A]): F[A] =
+    Sync[F].bracket(lockWrite)(_ => fa)(_ => unlockWrite)
+
+  // In case there's a write operation among the reads and we haven't
+  // taken out a write lock, we can upgrade the read lock for the duration
+  // of the write, then downgrade it back to read when we're done.
+  // See here for the rules up upgrading/downgrading:
+  // https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html
+  private def withLockUpgrade[A](fa: F[A]): F[A] =
```

Comment:
Appreciate you double checking the docs!

That was confusing, you got me there for a sec. I remember reading it then forgot because the code sample seems to be more or less doing the thing I wanted:

```scala
void processCachedData() {
     rwl.readLock().lock();
     if (!cacheValid) {
        // Must release read lock before acquiring write lock
        rwl.readLock().unlock();
        rwl.writeLock().lock();
        try {
          // Recheck state because another thread might have
          // acquired write lock and changed state before we did.
          if (!cacheValid) {
            data = ...
            cacheValid = true;
          }
          // Downgrade by acquiring read lock before releasing write lock
          rwl.readLock().lock();
        } finally {
          rwl.writeLock().unlock(); // Unlock write, still hold read
        }
     }

     try {
       use(data);
     } finally {
       rwl.readLock().unlock();
     }
   }
```

I think what they mean by "upgrading from a read lock to the write lock is not possible" is reflected here:
```scala
        // Must release read lock before acquiring write lock
        rwl.readLock().unlock();
        rwl.writeLock().lock();
```
You have to let go of your lock first, then acquire a write. But going to other way, you can hold both locks:
```scala
       // Downgrade by acquiring read lock before releasing write lock
          rwl.readLock().lock();
        } finally {
          rwl.writeLock().unlock(); // Unlock write, still hold read
        }
```
So indeed, technically it's not a real upgrade. That's why the method that uses this technique is meant for reading the database, but since I have no way to statically analyse the `Free` data structure to check if there's a write, all I can hope for is that the programmer chose the right version. In case they didn't, I'm falling back to this option where a temporary write lock is taken out in the middle, so at least it's not affecting other ongoing reads, but another write batch may go ahead before the write lock is acquired.

Reply:
then maybe I would change wording in docs, to avoid statement that we can upgrade read lock to write lock, which java docs say it is impossible.  This will only brings questions from the readers in the future.

Reply:
Done :+1: